File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.cpp
#include <string>
#include <vector>
#include <map>
#include <iostream>
#include <climits>
using namespace std;

vector<int> solution(vector<string> gems) {
    vector<int> answer;
    map<string,int> m;
    for(auto gem : gems)
        m[gem] = 0;
    
    int size = m.size();
    int start = 0;
    int min = INT_MAX;
    int cnt = 0;
    int x = 0, y = 0;
    for(int i = 0; i< gems.size(); i++){
        if(m[gems[i]] == 0)
            cnt++;
        m[gems[i]]++;
        
        if(m[gems[start]] > 1){
            while(m[gems[start]] > 1){
                m[gems[start]]--;
                start++;
            }
        }
        if(cnt == size && min > i - start + 1){
            min = i - start + 1;
            x = start + 1;
            y = i + 1;
        }
    }
    answer.push_back(x);
    answer.push_back(y);
    
    return answer;
}

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.java
import java.io.*;
import java.util.*;
class Solution {
    public int[] solution(String[] gems) {
        int[] answer = new int[2];
        
        Map<String, Integer> map = new HashMap<>();
        for(String s : gems) map.put(s, 0);
        
        int n = map.size();
        
        int x = 0;
        int start = 0;
        int a, b, len;
        a = b = len = Integer.MAX_VALUE;
        
        for(int i = 0; i < gems.length; i++){
            map.put(gems[i], map.get(gems[i]) + 1);
            
            if(map.get(gems[i]) == 1) x++;
            
            if(x < n) continue;
            
            while(start <= i && x == n){
                if(len > i - start){
                    len = i - start;
                    a = start;
                    b = i;
                }
                
                map.put(gems[start], map.get(gems[start]) - 1);
                if( map.get(gems[start]) == 0) x--;
                start++;
                
            }
            
            
        }
        answer[0] = a + 1;
        answer[1] = b + 1;
        
        return answer;
    }
}